pydicom/charset.py

```diff
@@ -57,68 +60,115 @@
 # Map Python encodings to escape sequences as defined in PS3.3 in tables
 # C.12-3 (single-byte) and C.12-4 (multi-byte character sets).
 escape_codes = {
-    'iso8859': ESC + b'(B',  # used to switch to ASCII G0 code element
-    'latin_1': ESC + b'-A',
-    'shift_jis': ESC + b')I',
-    'iso2022_jp': ESC + b'$B',
-    'iso8859_2': ESC + b'-B',
-    'iso8859_3': ESC + b'-C',
-    'iso8859_4': ESC + b'-D',
-    'iso_ir_126': ESC + b'-F',
-    'iso_ir_127': ESC + b'-G',
-    'iso_ir_138': ESC + b'-H',
-    'iso_ir_144': ESC + b'-L',
-    'iso_ir_148': ESC + b'-M',
-    'iso_ir_166': ESC + b'-T',
-    'euc_kr': ESC + b'$)C',
-    'iso-2022-jp': ESC + b'$(D',
-    'iso_ir_58': ESC + b'$)A',
+    ESC + b'(B': default_encoding,  # used to switch to ASCII G0 code element
+    ESC + b'-A': 'latin_1',
+    ESC + b')I': 'shift_jis',
+    ESC + b'$B': 'iso2022_jp',
+    ESC + b'-B': 'iso8859_2',
+    ESC + b'-C': 'iso8859_3',
+    ESC + b'-D': 'iso8859_4',
+    ESC + b'-F': 'iso_ir_126',
+    ESC + b'-G': 'iso_ir_127',
+    ESC + b'-H': 'iso_ir_138',
+    ESC + b'-L': 'iso_ir_144',
+    ESC + b'-M': 'iso_ir_148',
+    ESC + b'-T': 'iso_ir_166',
+    ESC + b'$)C': 'euc_kr',
+    ESC + b'$(D': 'iso-2022-jp',
+    ESC + b'$)A': 'iso_ir_58',
 }
 
-default_encoding = "iso8859"
-
 
-def decode_string(value, encodings):
+def decode_string(value, encodings, delims):
     """Convert a raw byte string into a unicode string using the given
     list of encodings.
+
+    Parameters
+    ----------
+    value : byte string
+        The raw string as encoded in the DICOM tag value.
+    encodings : list
+        The encodings needed to decode the string as a list of Python
+        encodings, converted from the encodings in Specific Character Set.
+    delims : byte string
+        A string containing all characters that may cause the encoding of
+        `value` to change.
+
+    Returns
+    -------
+    text type
+        The decoded string.
     """
     if ESC not in value:
         return value.decode(encodings[0])
 
     # multi-byte character sets except Korean are handled by Python encodings
-    use_python_handling = value.startswith((escape_codes['iso2022_jp'],
-                                            escape_codes['iso-2022-jp'],
-                                            escape_codes['iso_ir_58']))
+    use_python_handling = value.startswith((ESC + b'$B',  # 'iso2022_jp
+                                            ESC + b'$(D',  # iso-2022-jp
+                                            ESC + b'$)A'))  # iso_ir_58
 
```

Comment:
IIUC, this is only evaluated for the start of value, not for individual 'fragments'.  Is it possible for it to be True for the first encoding, but not on others?

Reply:
In this case I don't split the value currently, so the additional delimiters probably won't work with these encodings. As I wrote in another comment, this is still an open issue that I'm struggling with  (splitting breaks the decoding in these cases).

Reply:
Yes - this is what I wanted to do but couldn't get it to work. But it should work. I think I overdid the part with the delimiters - while it is described in the standard, that they should aways reset the encoding, I have no real example where this is the case. In the multi-byte encodings splitting by the delimiters won't work, because they may be part of a multi-byte character  (like one of the caret characters in Yamada Tarou, which actually is part of a kanji character), and I now think that in reality, they are always prepended by an escape sequence if needed, even if not always demanded by the standard as I understand it. 
Also, the splitting by '=' is still done in PN before the decoding, which is also wrong for multi-byte encodings. it just works accidentally with the example data.
I think I will clean this up at the weekend and remove the delimiter part at all.

Reply:
Thanks for your patience explaining all this ... I leave it to your judgement; my understanding of it is just barely scratching the surface.

Reply:
Well, I'm not much better off... 
You were right - I did the splitting for all escape characters, and it works now (turned out I just forgot an escape sequence - there are 2 for JISX). I still do the delimiter stuff for single byte encodings (though another way now, that I'm not too happy about), and I was also wrong about the '=' character - this was correct as it was.
Anyway, I will leave it at as is now, except for anything that comes up in the review, of course.